controllers/integration/integration_adapter.go

```diff
@@ -0,0 +1,310 @@
+func (a *Adapter) compareApplicationSnapshots(expectedApplicationSnapshot *releasev1alpha1.ApplicationSnapshot, foundApplicationSnapshot *releasev1alpha1.ApplicationSnapshot) bool {
+	allImagesMatch := true
+	for _, image1 := range expectedApplicationSnapshot.Spec.Images {
+		foundImage := false
+		for _, image2 := range foundApplicationSnapshot.Spec.Images {
+			if image2 == image1 {
+				foundImage = true
+			}
+		}
+		if !foundImage {
+			allImagesMatch = false
+		}
+
+	}
+	return allImagesMatch
+}
+
+// findMatchingApplicationSnapshot tries to find the expected ApplicationSnapshot with the same set of images
+func (a *Adapter) findMatchingApplicationSnapshot() (*releasev1alpha1.ApplicationSnapshot, error) {
+	var allApplicationSnapshots *[]releasev1alpha1.ApplicationSnapshot
+	expectedApplicationSnapshot, err := a.createApplicationSnapshotForPipelineRun()
+	if err != nil {
+		return nil, err
+	}
+	allApplicationSnapshots, err = a.getAllApplicationSnapshots()
+	if err != nil {
+		return nil, err
+	}
+	for _, foundApplicationSnapshot := range *allApplicationSnapshots {
+		if a.compareApplicationSnapshots(expectedApplicationSnapshot, &foundApplicationSnapshot) {
+			return &foundApplicationSnapshot, nil
+		}
+	}
+	return nil, nil
+}
+
+// getAllApplicationComponents loads from the cluster all Components associated with the given Application.
+// If the Application doesn't have any Components or this is not found in the cluster, an error will be returned.
+func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Application) (*[]hasv1alpha1.Component, error) {
+	applicationComponents := &hasv1alpha1.ComponentList{}
+	opts := []client.ListOption{
+		client.InNamespace(application.Namespace),
+		client.MatchingFields{"spec.application": application.Name},
+	}
+
+	err := a.client.List(a.context, applicationComponents, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationComponents.Items, nil
+}
+
+// getAllApplicationReleaseLinks returns the ReleaseLinks used by the application being processed. If matching
+// ReleaseLinks are not found, an error will be returned.
+func (a *Adapter) getAllApplicationReleaseLinks() (*[]releasev1alpha1.ReleaseLink, error) {
+	releaseLinks := &releasev1alpha1.ReleaseLinkList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+	}
+
+	err := a.client.List(a.context, releaseLinks, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releaseLinks.Items, nil
+}
+
+// getApplicationSnapshot returns the all ApplicationSnapshots in the Application's namespace nil if it's not found.
+// In the case the List operation fails, an error will be returned.
+func (a *Adapter) getAllApplicationSnapshotReleases(applicationSnapshot *releasev1alpha1.ApplicationSnapshot) (*[]releasev1alpha1.Release, error) {
```

Comment:
I'd rename it to `getReleasesWithApplicationSnapshot`. I think it's easier to understand what it does and it adheres to Go conventions.